File: src/py_pkg/py_pkg/smartphone.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import String


class SmartphoneNode(Node):
    def __init__(self):
        super().__init__("smartphone")
        self.subscriber=self.create_subscription(String,"news",self.robot_callback,10)
        self.get_logger().info("smart phone has been started")
        self.message_count = 0

    def robot_callback(self, msg: String):
        self.message_count += 1
        self.get_logger().info(f"[消息 #{self.message_count}] {msg.data}")
        self.get_logger().info(f"[消息 #{self.message_count}] {msg.data}")
        self.get_logger().info(f"[消息 #{self.message_count}] {msg.data}")
        self.get_logger().info(f"[消息 #{self.message_count}] {msg.data}")
        self.get_logger().info(f"[消息 #{self.message_count}] {msg.data}")
def main(args=None):
    rclpy.init(args=args)
    node = SmartphoneNode()
    rclpy.spin(node)
    rclpy.shutdown()

if __name__=="__main__":
    main()

File: ros2_level1-2/src/cpp_pkg/src/smartphone.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/string.hpp"




class Smartphone: public rclcpp::Node
{
  public:
  Smartphone():Node("smartphone")
  {
    subscriber_=this->create_subscription<example_interfaces::msg::String>("smartphone_news",10,
      std::bind(&Smartphone::callbacknews,this,std::placeholders::_1));
      RCLCPP_INFO(this->get_logger(), "Smartphone node has been started.");

  }
private:
void callbacknews(const example_interfaces::msg::String::SharedPtr msg){
    RCLCPP_INFO(this->get_logger(),"news: %s",msg->data.c_str());

}
rclcpp::Subscription<example_interfaces::msg::String>::SharedPtr subscriber_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node=std::make_shared<Smartphone>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
